pkg/reconciler/namespace/controller.go

```diff
@@ -0,0 +1,256 @@
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+
+	"github.com/kcp-dev/kcp/pkg/apis/cluster/v1alpha1"
+	clusterclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	"github.com/kcp-dev/kcp/pkg/client/informers/externalversions"
+	clusterlisters "github.com/kcp-dev/kcp/pkg/client/listers/cluster/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/gvk"
+	"github.com/kcp-dev/kcp/pkg/informer"
+)
+
+const resyncPeriod = 10 * time.Hour
+
+// NewController returns a new Controller which schedules namespaced resources to a Cluster.
+func NewController(cfg *rest.Config) *Controller {
+	dynClient := dynamic.NewForConfigOrDie(cfg)
+	kubeClient := kubernetes.NewForConfigOrDie(cfg)
+	clusterClient := clusterclient.NewForConfigOrDie(cfg)
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	stopCh := make(chan struct{}) // TODO: hook this up to SIGTERM/SIGINT
+
+	c := &Controller{
+		queue:      queue,
+		dynClient:  dynClient,
+		kubeClient: kubeClient,
+		stopCh:     stopCh,
+		gvkTrans:   gvk.NewGVKTranslator(cfg),
+	}
+
+	ddsif := informer.NewDynamicDiscoverySharedInformerFactory(cfg, nil, informer.GVREventHandlerFuncs{
+		AddFunc:    func(gvr schema.GroupVersionResource, obj interface{}) { c.enqueueResource(gvr, obj) },
+		UpdateFunc: func(gvr schema.GroupVersionResource, _, obj interface{}) { c.enqueueResource(gvr, obj) },
+		DeleteFunc: func(gvr schema.GroupVersionResource, obj interface{}) { c.enqueueResource(gvr, obj) },
+	})
+	ddsif.WaitForCacheSync(stopCh)
+	ddsif.Start(stopCh)
+
+	csif := externalversions.NewSharedInformerFactoryWithOptions(clusterClient, resyncPeriod)
+	csif.Cluster().V1alpha1().Clusters().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueCluster(obj) },
+		UpdateFunc: func(_, obj interface{}) { c.enqueueCluster(obj) },
+		DeleteFunc: func(obj interface{}) { c.enqueueCluster(obj) },
+	})
+	c.clusterLister = csif.Cluster().V1alpha1().Clusters().Lister()
+	csif.WaitForCacheSync(stopCh)
+	csif.Start(stopCh)
+
+	nsif := informers.NewSharedInformerFactoryWithOptions(kubeClient, resyncPeriod)
+	nsif.Core().V1().Namespaces().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueNamespace(obj) },
+		UpdateFunc: func(_, obj interface{}) { c.enqueueNamespace(obj) },
+		DeleteFunc: func(obj interface{}) { c.enqueueNamespace(obj) },
+	})
+	c.namespaceLister = nsif.Core().V1().Namespaces().Lister()
+	nsif.WaitForCacheSync(stopCh)
+	nsif.Start(stopCh)
+
+	return c
+}
+
+type Controller struct {
+	queue workqueue.RateLimitingInterface
+
+	dynClient  dynamic.Interface
+	kubeClient kubernetes.Interface
+
+	clusterLister   clusterlisters.ClusterLister
+	namespaceLister corelister.NamespaceLister
+	namespaceClient coreclient.NamespaceInterface
+	stopCh          chan struct{}
+	indexer         cache.Indexer
+	lister          cache.GenericLister
```

Comment:
Does this ever get initialized?

Reply:
It wasn't. I ended up replacing the `lister` and `indexer` both with somewhat terrible hacks to expose the per-type listers and indexers from the dynamic-discovery SIF. I'll play with this some more today and see what breaks.